SignalServiceKit/src/Devices/OWSDevice.m

```diff
@@ -49,30 +46,34 @@ - (BOOL)mayHaveLinkedDevices:(YapDatabaseConnection *)dbConnection
 {
     OWSAssert(dbConnection);
 
-    @synchronized(self)
-    {
-        if (!self.mayHaveLinkedDevicesCached) {
-            self.mayHaveLinkedDevicesCached = @([dbConnection boolForKey:kTSStorageManager_MayHaveLinkedDevices
-                                                            inCollection:kTSStorageManager_OWSDeviceCollection
-                                                            defaultValue:YES]);
-        }
+    return [dbConnection boolForKey:kTSStorageManager_MayHaveLinkedDevices
+                       inCollection:kTSStorageManager_OWSDeviceCollection
+                       defaultValue:YES];
+}
 
-        return [self.mayHaveLinkedDevicesCached boolValue];
-    }
+// In order to avoid skipping necessary sync messages, the default value
+// for mayHaveLinkedDevices is YES.  Once we've successfully sent a
+// sync message with no device messages (e.g. the service has confirmed
+// that we have no linked devices), we can set mayHaveLinkedDevices to NO
+// to avoid unnecessary message sends for sync messages until we learn
+// of a linked device (e.g. through the device linking UI or by receiving
+// a sync message, etc.).
+- (void)clearMayHaveLinkedDevicesIfNotSet
```

Comment:
But as soon as we know that these sends aren't necessary, we should start skipping them... until we have new reason they might be necessary.